pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/RangeIndexReader.java

```diff
@@ -0,0 +1,207 @@
+        }
+        break;
+      case FLOAT:
+        for (int i = 0; i < _numRanges; i++) {
+          _rangeStartArray[i] = _buffer.getFloat(rangeArrayStartOffset + i * Float.BYTES);
+        }
+        break;
+      case DOUBLE:
+        for (int i = 0; i < _numRanges; i++) {
+          _rangeStartArray[i] = _buffer.getDouble(rangeArrayStartOffset + i * Double.BYTES);
+        }
+        break;
+    }
+  }
+
+  @VisibleForTesting
+  public Number[] getRangeStartArray() {
+    return _rangeStartArray;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public ImmutableRoaringBitmap getDocIds(int rangeId) {
+    SoftReference<ImmutableRoaringBitmap>[] bitmapArrayReference = null;
+    // Return the bitmap if it's still on heap
+    if (_bitmaps != null) {
+      bitmapArrayReference = _bitmaps.get();
+      if (bitmapArrayReference != null) {
+        SoftReference<ImmutableRoaringBitmap> bitmapReference = bitmapArrayReference[rangeId];
+        if (bitmapReference != null) {
+          ImmutableRoaringBitmap value = bitmapReference.get();
+          if (value != null) {
+            return value;
+          }
+        }
+      } else {
+        bitmapArrayReference = new SoftReference[_numRanges];
+        _bitmaps = new SoftReference<SoftReference<ImmutableRoaringBitmap>[]>(bitmapArrayReference);
+      }
+    } else {
+      bitmapArrayReference = new SoftReference[_numRanges];
+      _bitmaps = new SoftReference<SoftReference<ImmutableRoaringBitmap>[]>(bitmapArrayReference);
+    }
+    synchronized (this) {
+      ImmutableRoaringBitmap value;
+      if (bitmapArrayReference[rangeId] == null || bitmapArrayReference[rangeId].get() == null) {
+        value = buildRoaringBitmapForIndex(rangeId);
+        bitmapArrayReference[rangeId] = new SoftReference<ImmutableRoaringBitmap>(value);
+      } else {
+        value = bitmapArrayReference[rangeId].get();
+      }
+      return value;
+    }
+  }
+
+  @Override
+  public ImmutableRoaringBitmap getDocIds(Object value) {
+    // This should not be called from anywhere. If it happens, there is a bug
+    // and that's why we throw illegal state exception
+    throw new IllegalStateException("bitmap inverted index reader supports lookup only on dictionary id");
+  }
+
+  private synchronized ImmutableRoaringBitmap buildRoaringBitmapForIndex(final int rangeId) {
+    final long currentOffset = getOffset(rangeId);
+    final long nextOffset = getOffset(rangeId + 1);
+    final int bufferLength = (int) (nextOffset - currentOffset);
+
+    // Slice the buffer appropriately for Roaring Bitmap
+    ByteBuffer bb = _buffer.toDirectByteBuffer(currentOffset, bufferLength);
+    return new ImmutableRoaringBitmap(bb);
+  }
+
+  private long getOffset(final int rangeId) {
+    return _buffer.getLong(_bitmapIndexOffset + rangeId * Long.BYTES);
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    _buffer.close();
+  }
+
+  public int findRangeId(int value) {
```

Comment:
Can we add a TODO here to indicate that for now we are doing a linear scan to find the corresponding range for a given dictId. If the number of ranges is high, we may want to do binary search?